### ISSUE TITLE
L045 should handle nested CTEs

### ISSUE THREAD

**Issue by barrywhart** [sqlfluff/sqlfluff#1999]
(I plan to work on this myself, so this is a very abbreviated issue description.)

L045 shares much of its implementation with L044. Both use `SelectCrawler`, and they were not originally designed to handle nested CTEs.

Reference these similar things for L044:
* Issue #1984
* PR #1991

## Version
Include the output of `sqlfluff --version` along with your Python version

## Configuration
```
Include your SQLFluff configuration here
```

**Comment by kdw2126** [comment]
Just to provide a simple test case of how L045 can currently fail:

```sql
WITH
    foo AS (
        SELECT
            *
        FROM
            zipcode
    ),

    bar AS (
        SELECT
            *
        FROM
            county
    ),

    stage AS (
        (SELECT
            *
        FROM
            foo)

            UNION ALL

        (SELECT
            *
        FROM
            bar)
    )

    SELECT
        *
    FROM
        stage
```

You get a bunch of L045 errors for this sample query.

**Comment by barrywhart** [comment]
Linting this query should report that CTE `b` is unused.

```
with a as (
    with b as (
        select 1 from foo
    )
    select 1
)
select * from a
```

**Comment by barrywhart** [comment]
The test query provided by @kdw2126 is for a slightly different L045 issue -- nested queries vs nested CTEs. I'll go ahead and fix both in this issue.
